projects/train/train/waveform_sampler.py

```diff
@@ -0,0 +1,254 @@
+        if self.path is not None:
+            for path in self.path:
+                group = group[path]
+        return f, group
+
+    def sample_fnames(self) -> np.ndarray:
+        return np.random.choice(
+            self.fnames,
+            p=self.probs,
+            size=(self.chunks_per_batch,),
+            replace=True,
+        )
+
+    def load_chunk(self, fname, start, size):
+        end = min(start + size, self.sizes[fname])
+        return {
+            channel: self.mmap_datasets[fname][channel][start:end]
+            for channel in self.channels
+        }
+
+    def sample_batch(self):
+        # allocate batch up front
+        batch = np.zeros(
+            (self.batch_size, self.num_channels, self.waveform_size)
+        )
+
+        for i in range(self.chunks_per_batch):
+            fname = np.random.choice(self.fnames, p=self.probs)
+
+            chunk_size = min(
+                self.chunk_size, self.batch_size - i * self.chunk_size
+            )
+
+            # select a random starting index for the chunk
+            max_start = self.sizes[fname] - chunk_size
+            start = np.random.randint(0, max_start + 1)
+
+            # load the chunk and insert it into the batch
+            chunk = self.load_chunk(fname, start, chunk_size)
+            batch_start = i * self.chunk_size
+            batch_end = batch_start + chunk_size
+
+            for i, channel in enumerate(self.channels):
+                batch[batch_start:batch_end, i, :] = chunk[channel]
+
+        return torch.tensor(batch)
+
+    def __iter__(self):
+        for _ in range(self.batches_per_epoch):
+            yield self.sample_batch()
+
+
+class ChunkedWaveformDataset(torch.utils.data.IterableDataset):
+    """
+    Wrapper dataset that will loop through chunks of timeseries
+    data produced by another iterable and sample subsets
+    of waveforms from each chunk.
+
+    Args:
+        chunk_it:
+            Iterator which will produce batches of waveform
+            data to sample subsets from. Should have shape
+            `(N, C, T)`, where `N` is the number of waveformns
+            to sample from, `C` is the number of channels,
+            and `T` is the number of samples along the
+            time dimension for each waveform.
+        batch_size:
+            Number of waveforms to sample at each iteration
+        batches_per_chunk:
+            Number of batches of waveforms to sample from
+            each chunk before moving on to the next one.
+    """
+
+    def __init__(
+        self,
+        chunk_it: Iterable,
+        batch_size: int,
+        batches_per_chunk: int,
+    ) -> None:
+        self.logger = logging.getLogger(__name__)
+        self.chunk_it = chunk_it
+        self.batch_size = batch_size
+        self.batches_per_chunk = batches_per_chunk
+
+    def __len__(self):
+        return len(self.chunk_it) * self.batches_per_chunk
+
+    def __iter__(self):
+        it = iter(self.chunk_it)
```

Comment:
Is this casting necessary?

Reply:
I think so - need to initialize the `chunk_it` as an iterator so that we can call `next` on it down the line.